plugins/newrelic/src/api/index.ts

```diff
@@ -71,39 +76,105 @@ export interface NewRelicApi {
   getApplications(): Promise<NewRelicApplications>;
 }
 
+interface PaginationInformation {
+  nextPageLink: string;
+  relation: string;
+}
+
+interface NewRelicPageReadResult {
+  hasMoreApplications: boolean;
+  pagination: PaginationInformation | undefined;
+  applicationsFromReadPage: NewRelicApplication[];
+}
+
 export class NewRelicClient implements NewRelicApi {
   private readonly discoveryApi: DiscoveryApi;
+  private readonly fetchApi: FetchApi;
   private readonly proxyPathBase: string;
+  private baseUrl: string;
 
   constructor(options: Options) {
     this.discoveryApi = options.discoveryApi;
+    this.fetchApi = options.fetchApi;
     this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
+    this.baseUrl = '';
   }
 
   async getApplications(): Promise<NewRelicApplications> {
-    const url = await this.getApiUrl('apm', 'applications.json');
-    const response = await fetch(url);
-    let responseJson;
+    if (!this.baseUrl) {
+      const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
+      this.baseUrl = `${proxyUrl}${this.proxyPathBase}/apm/api/applications.json`;
+    }
+
+    let targetUrl = this.baseUrl;
+    let hasNextPage = true;
 
     try {
-      responseJson = await response.json();
+      let applications: NewRelicApplication[] = [];
+
+      do {
+        const { hasMoreApplications, pagination, applicationsFromReadPage } =
+          await this.fetchNewRelic(targetUrl);
+
+        hasNextPage = hasMoreApplications;
+        targetUrl = hasNextPage ? pagination!.nextPageLink : '';
+        applications = applications.concat(applicationsFromReadPage);
+      } while (hasNextPage);
+
+      return { applications };
     } catch (e) {
-      responseJson = { applications: [] };
+      return { applications: [] };
     }
+  }
+
+  private async fetchNewRelic(
+    targetUrl: string,
+  ): Promise<NewRelicPageReadResult> {
+    const response = await this.fetchApi.fetch(targetUrl);
+    const responseJson = await response.json();
 
-    if (response.status !== 200) {
+    if (!response.ok) {
```

Comment:
you need to do this check before the `.json` call, since that call also checks `ok` and throws a generic error if it isn't

Reply:
Is the specific behavior of ```.json``` defined by the browser? I have not observed this behavior with fetch, but I use Chrome. Wouldn't checking the ```ok``` and throwing a generic error discard potentially important information from the server if the error does not also include the body of the response?

To test this, I have thrown together a quick C# app that just creates an ```HttpListener``` that returns a status code of 500 with a body of ```{"message": "TESTING"}```. A created a route to it through the proxy called ```/testingBreakingThings```.

I also modified the code to fetch that route instead and do so outside of any ```try/catch``` blocks, like so:
![image](https://github.com/backstage/backstage/assets/129326788/f638edf6-399b-4b04-9be6-75977bd2cd35)

Despite these changes, it is able to get the json response without any errors thrown, and the if statement here takes over and throws the "Error communicating with New Relic..." error:
![image](https://github.com/backstage/backstage/assets/129326788/6152128a-4a4f-4766-b77a-b528be2adede)

It also behaves this way through the browser console:
![image](https://github.com/backstage/backstage/assets/129326788/4d0a53e1-899a-4f6d-8e1e-f42b133cd00c)

Reply:
Oh this is just me being confused and not very clear. I've rather been bitten by the `.json` throwing when the response _wasn't_ json shaped, which kind of makes the errors get mixed up. Your error will be like `Illegal character < in ...` or whatnot if it was instead an HTML page returned by a rate limiter or ingress proxy for example. In this case if the body _is_ json shaped and parses correctly it does work in our favour.

Reply:
Ah, yes. If the response body is not JSON then the parse will throw an error like what you have described. I thought that I had been replicating the plugin's previous error handling behavior, but I believe I was mistaken with what the behavior was the first time I looked at it. 

The previous error handling behavior was:
- Check the response status code first, throw an error if it is not ok
- Attempt to parse the JSON, catching any errors. If there is a parse error, return an empty array of applications.

Would we maybe want to change the behavior to something like this?
- Check the response status code first, throw an error if it is not ok
- Parse the JSON outside of any try/catch, so in case the response is ok but not JSON any thrown errors escape and get displayed (instead of the table just being empty)

Either way, I am going to switch to checking if it is ok first so http errors surface first 👍

Reply:
Sure!